pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/PinotHelixResourceManager.java

```diff
@@ -3148,6 +3137,31 @@ public PinotResourceManagerResponse dropInstance(String instanceName) {
     return PinotResourceManagerResponse.success("Instance " + instanceName + " dropped");
   }
 
+  /**
+   * Utility to perform a safety check of the operation to drop an instance.
+   * If the resource is not safe to drop the utility lists all the possible reasons.
+   * @param instanceName Pinot instance name
+   * @return List of reasons why the operation is unsafe. Empty list if the operation is safe.
+   */
+  public List<String> instanceDropSafetyCheck(String instanceName) {
+    List<String> issues = new ArrayList<>();
+    // Check if the instance is live
+    if (_helixDataAccessor.getProperty(_keyBuilder.liveInstance(instanceName)) != null) {
+      issues.add(String.format("Instance %s is still live", instanceName));
+    }
+    // Check if any ideal state includes the instance
+    getAllResources().forEach(resource -> {
+      IdealState idealState = _helixAdmin.getResourceIdealState(_helixClusterName, resource);
+      for (String partition : idealState.getPartitionSet()) {
+        if (idealState.getInstanceSet(partition).contains(instanceName)) {
+          issues.add(String.format("Instance %s exists in ideal state for %s", instanceName, resource));
+          break;
```

Comment:
We're early terminating here, but not early terminating if the instance is live? Is this intentional? Or should we include all tables where this instance is part of the ideal state to give a comprehensive list of issues?

Reply:
We are just early terminating the loop on resource partitions, like even if one segment is present on the instance mark it as unsafe saying that particular table is on that instance. The outer loop will still check for all the tables and give a comprehensive list of issues at table level.